### ISSUE TITLE
Allow `transform(df, [:x1, Between(:x2, :x4)] => fun)`

### ISSUE THREAD

**Issue by pdeffebach** [JuliaData/DataFrames.jl#2328]
```
df = DataFrame(rand(10, 4))
transform(df, [:x1, Between(:x2, :x4)] => fun)
```

where `fun` accepts a vector and a vector of vectors. i.e., when multi-column selectors are used inside the function input vector, default to a vector of vectors.

**Comment by bkamins** [comment]
In order to support this we would have to define a set of rules how nesting in general should be resolved. Currently nesting is not allowed anywhere except `All`.

In general the problem with this proposal, is that such a call in general would not be type stable, and currently the rules we have guarantee type stability of the call.

**Comment by pdeffebach** [comment]
Yes. I wonder if we should allow `AsTable` to be an exception in the `Cols` / `All`. 

Then you could input a named tuple into the function.

**Comment by pdeffebach** [comment]
For other readers, this is a selfish issue referenced in [this issue](https://github.com/JuliaData/DataFramesMeta.jl/issues/158) in DataFramesMeta.

**Comment by bkamins** [comment]
In general - before moving forward we need a grammar here. The current design has a precise grammar that things are allowed if and only if they are a valid index into a data frame, except for duplicates, which are allowed (and are disallowed in indexing).

**Comment by pdeffebach** [comment]
Could I write out documentation as a start? I think special-casing `AsTable` to be included as function arguments is feasible.

**Comment by bkamins** [comment]
Maybe please just first write down the rules you would like here as a comment so that we are clear what level of sophistication you would like to see.

In general in the long run we should refactor `combine`/`select`/`transform` to follow the pattern with common components:
1. common preprocessing
2. common transformation
3. common postprocessing

Then your proposal would go into the common preprocessing.
(other things that would also go in then are e.g. allowing returning tables from functions in `select`/`transform`)

The current design works but is very complex (this is a typical consequence of gradual adding of functionalities)

**Comment by pdeffebach** [comment]
Okay my specific proposal is to allow `AsTable` objects in the inputs to a function in `select` and `transform`. 

```
function fun(x::Vector{Real}, y::NamedTupleofVectors)
    sum(y) ./ x
end

income_vars = [:non_farm_incom, :farm_income, :rent_income]

transform(df, [:price, AsTable(income_vars)] => fun => :adjusted_income)
```

Perhaps as @nalimilan [mentioned](https://github.com/JuliaData/DataFrames.jl/issues/2325#issuecomment-664436330) in #2325 we could also include `RowNumber()` as something that could be inputted along with other variables into a function.

**Comment by bkamins** [comment]
Just to make the proposal more precise, if I understand what you propose correctly:

`AsTable` and `RowNumber` should be allowed as entries when a vector is passed as a first argument in `src => fun => dst`. Is this correct?

**Comment by pdeffebach** [comment]
Yes, exactly.

**Comment by pdeffebach** [comment]
Additionally, `NRow` should be it's own singleton struct in this context.

**Comment by bkamins** [comment]
What would `NRow` be? Total number of rows in the source data frame?

**Comment by pdeffebach** [comment]
yes. In `dplyr` you can do 

```
df2 = df %>%
   group_by(groupvar) %>%
   mutate(y = x / n())
```

Of course, `length` of some arbitrary variable is also fine to use. This is not a super necessary addition.

**Comment by bkamins** [comment]
Then this is something as that can do into DataFramesMeta.jl but not DataFrames.jl. What syntax for DataFrames.jl would you envision for it, as normally I would just write `transform(df, :x => (x -> x / nrow(df)) => :y)` now to have what you described.

**Comment by nalimilan** [comment]
In DataFrames that would be e.g. `[:x, NRow] => ((x, n) -> x / n) => :y`. That would be useful for type-stable operations on `GroupedDataFrame`. Note that the `nrow` function could be enough, maybe `NRow` isn't needed.

**Comment by bkamins** [comment]
In `GroupedDataFrame` you would write:
```
transform(gdf, :x => (x -> x / length(x)) => :y)
```
The only place where I would see the benefit of it is in `ByRow`, where stand-alone you do not have access to the length.

If we wanted to go for it then `NRow` could be `nrow` or just `length` and `RowNumber` could be just `eachindex` in order to avoid introducing new names.
